workloads/librispeech/librispeech_jax/submission.py

```diff
@@ -0,0 +1,156 @@
+
+
+# We need to jax.pmap here instead of inside update_params because the latter
+# would recompile the function every step.
+@functools.partial(
+  jax.pmap,
+  axis_name='batch',
+  in_axes=(None, None, 0, 0, 0, None, 0, None),
+  static_broadcasted_argnums=(0, 1))
+def pmapped_train_step(workload, opt_update_fn, model_state, optimizer_state,
+                       current_param_container, hyperparameters, batch, rng):
+  def _loss_fn(params):
+    """loss function used for training."""
+    variables = {'params': params, **model_state}
+    logits, new_model_state = workload.model_fn(
+        params,
+        batch['input'],
+        model_state,
+        spec.ForwardPassMode.TRAIN,
+        rng,
+        update_batch_norm=True)
+    loss = workload.loss_fn(batch['label'], logits)
+    return loss, (new_model_state, logits)
+
+  grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+  aux, grad = grad_fn(current_param_container)
+  grad = lax.pmean(grad, axis_name='batch')
+  new_model_state, logits = aux[1]
+  updates, new_optimizer_state = opt_update_fn(
+      grad, optimizer_state, current_param_container)
+  updated_params = optax.apply_updates(current_param_container, updates)
+
+  return new_model_state, new_optimizer_state, updated_params
+
+def update_params(
+    workload: spec.Workload,
+    current_param_container: spec.ParameterContainer,
+    current_params_types: spec.ParameterTypeTree,
+    model_state: spec.ModelAuxiliaryState,
+    hyperparameters: spec.Hyperparamters,
+    input_batch: spec.Tensor,
+    label_batch: spec.Tensor,
+    loss_type: spec.LossType,
+    optimizer_state: spec.OptimizerState,
+    eval_results: List[Tuple[int, float]],
+    global_step: int,
+    rng: spec.RandomState) -> spec.UpdateReturn:
+  """Return (updated_optimizer_state, updated_params, updated_model_state)."""
+  _, features, transcripts, input_lengths,  transcripts_padding = input_batch
+  features = jnp.expand_dims(features.transpose(0, 2, 1), axis=1)
+  num_devices = jax.local_device_count()
+  reshaped_features = jnp.reshape(
+      features,
+      (num_devices, features.shape[0] // num_devices, *features.shape[1:]))
+  reshaped_input_lengths = jnp.reshape(
+      input_lengths,
+      (num_devices, input_lengths.shape[0] // num_devices, *input_lengths.shape[1:]))
+  reshaped_transcripts = jnp.reshape(
+      transcripts,
+      (num_devices, transcripts.shape[0] // num_devices,
+       *transcripts.shape[1:]))
+  reshaped_transcripts_padding = jnp.reshape(
+      transcripts_padding,
+      (num_devices, transcripts_padding.shape[0] // num_devices,
+       *transcripts_padding.shape[1:]))
+  sequential = [
+      nn.Conv(
+        features=32,
+        kernel_size=(41, 11),
+        strides=(2, 2),
+        padding=((20, 20), (5, 5)),
+      ),
+      nn.Conv(
+        features=32,
+        kernel_size=(21, 11),
+        strides=(2, 1),
+        padding=((10, 10), (5, 5)),
+      )
+    ]
```

Comment:
do we need to write out nn.Conv modules for this? can't we just provide the following to the function (hardcoded) instead of needing to import nn in this file?
- padding[1][1]
- kernel_dilation,
- kernel_size[1]
- strides[1]